contracts/tunnel-consumer/src/ibc.rs

```diff
@@ -0,0 +1,159 @@
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::entry_point;
+use cosmwasm_std::{
+    attr, from_json, DepsMut, Env, Ibc3ChannelOpenResponse, IbcBasicResponse, IbcChannel,
+    IbcChannelCloseMsg, IbcChannelConnectMsg, IbcChannelOpenMsg, IbcChannelOpenResponse, IbcPacket,
+    IbcPacketAckMsg, IbcPacketReceiveMsg, IbcPacketTimeoutMsg, IbcReceiveResponse, Never, StdAck,
+};
+
+use cw_band::tunnel::packet::{ack_fail, ack_success, TunnelPacket};
+use cw_band::tunnel::{TUNNEL_APP_VERSION, TUNNEL_ORDER};
+
+use crate::state::{SIGNAL_PRICE, TUNNEL_CONFIG};
+use crate::ContractError;
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn ibc_channel_open(
+    _deps: DepsMut,
+    _env: Env,
+    msg: IbcChannelOpenMsg,
+) -> Result<IbcChannelOpenResponse, ContractError> {
+    enforce_order_and_version(msg.channel(), msg.counterparty_version())?;
+
+    Ok(Some(Ibc3ChannelOpenResponse {
+        version: msg.channel().version.clone(),
+    }))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn ibc_channel_connect(
+    _deps: DepsMut,
+    _env: Env,
+    msg: IbcChannelConnectMsg,
+) -> Result<IbcBasicResponse, ContractError> {
+    enforce_order_and_version(msg.channel(), msg.counterparty_version())?;
+
+    Ok(IbcBasicResponse::default())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn ibc_channel_close(
+    _deps: DepsMut,
+    _env: Env,
+    _channel: IbcChannelCloseMsg,
+) -> Result<IbcBasicResponse, ContractError> {
+    Ok(IbcBasicResponse::default())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn ibc_packet_receive(
+    deps: DepsMut,
+    _env: Env,
+    msg: IbcPacketReceiveMsg,
+) -> Result<IbcReceiveResponse, Never> {
+    let packet = msg.packet;
+
+    do_ibc_packet_receive(deps, &packet).or_else(|err| {
+        Ok(
+            IbcReceiveResponse::new(ack_fail(err.to_string())).add_attributes(vec![
+                attr("action", "receive"),
+                attr("success", "false"),
+                attr("error", err.to_string()),
+            ]),
+        )
+    })
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn ibc_packet_ack(
+    _deps: DepsMut,
+    _env: Env,
+    msg: IbcPacketAckMsg,
+) -> Result<IbcBasicResponse, ContractError> {
+    let std_ack: StdAck = from_json(&msg.acknowledgement.data)?;
```

Comment:
```suggestion
    let std_ack = from_json(&msg.acknowledgement.data)?;
```

Reply:
its better to keep type hinting here since from_json returns T